### PR TITLE
Impl rewrite_result for ast nodes in items.rs

### PR DESCRIPTION
**Pull request by ding-young**
Tracked by #6206 

## Description 
- added `RewriteError` and blanket implementation for `rewrite_result` 
- above 6 ast nodes that implements rewrite, replaced rewrite function body with `rewrite_result` for 4 nodes; 
**ast::Local, ast::FieldDef, ast::Param, ast::FnRetTy**

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -2072,20 +2111,40 @@ impl<'a> Rewrite for OpaqueType<'a> {
...
+                ty.rewrite_result(
+                    context,
+                    shape
+                        .offset_left(3)
+                        .ok_or_else(|| RewriteError::ExceedsMaxWidth {
+                            configured_width: shape.width,
+                            span: self.span(),
+                        })?,
+                )
+                .map(|s| format!("-> {}", s))
```
Comment (ytmimi) [inline src/items.rs:2147]: To reduce some of the nesting it might make sense to refactor this slightly:

```rust
// 3 = "-> "
let shape = shape.offset_left(3).ok_or_else(|| RewriteError::ExceedsMaxWidth {
    configured_width: shape.width,
    span: self.span(),
})?;

ty.rewrite_result(context, shape).map(|s| format!("-> {}", s))
```

File: src/items.rs [hunk 2]
```diff
@@ -2072,20 +2111,40 @@ impl<'a> Rewrite for OpaqueType<'a> {
...
                 if context.config.version() == Version::One
                     || context.config.indent_style() == IndentStyle::Visual
                 {
-                    let inner_width = shape.width.checked_sub(3)?;
+                    let inner_width = shape.width.checked_sub(3).ok_or_else(|| {
+                        RewriteError::ExceedsMaxWidth {
+                            configured_width: shape.width,
+                            span: self.span(),
+                        }
+                    })?;
```
Comment (ytmimi) [inline src/items.rs:2132]: This isn't something we need to address right now, but I've noticed a lot of `.ok_or_else(|| { RewriteError::ExceedsMaxWidth {..} })` calls throughout the code.

I wonder if we can make this a little more ergonomic. Here are three ideas I have:
1. Define new helper methods on `Shape`
2. Add a macro that will handle some of this boilerplate
3. Define an extension trait so we can more easily convert from `Option<T>` -> `Result<T, RewriteError>`.
   Here's one idea that I had:

   ```rust
   /// Extension trait used to conveniently convert to RewriteError
   pub(crate) trait RewriteErrorExtension<T> {
       fn max_with_error(self, width: usize, span: Span) -> Result<T, RewriteError>;
       fn unknown_error(self) -> Result<T, RewriteError>;
   }

   impl<T> RewriteErrorExtension<T> for Option<T> {
       fn max_with_error(self, width: usize, span: Span) -> Result<T, RewriteError> {
           self.ok_or_else(|| RewriteError::ExceedsMaxWidth {
               configured_width: width,
               span: span,
           })
       }

       fn unknown_error(self) -> Result<T, RewriteError> {
           self.ok_or_else(|| RewriteError::Unknown)
       }
   }
   ```

File: src/items.rs [hunk 3]
```diff
@@ -2161,18 +2228,20 @@ impl Rewrite for ast::Param {
                 shape,
                 has_multiple_attr_lines,
             )
+            .ok_or_else(|| RewriteError::Unknown)
```
Comment (ytmimi) [inline src/items.rs:2231]: might also want to see how difficult it would be to change the signature of `rewrite_explicit_self`

File: src/items.rs [hunk 4]
```diff
@@ -2161,18 +2228,20 @@ impl Rewrite for ast::Param {
...
             let mut result = combine_strs_with_missing_comments(
                 context,
                 &param_attrs_result,
                 param_name,
                 span,
                 shape,
                 !has_multiple_attr_lines && !has_doc_comments,
-            )?;
+            )
+            .ok_or_else(|| RewriteError::Unknown)?;
```
Comment (ytmimi) [inline src/items.rs:2244]: Might make sense to keep track that we'll eventually want to change the signature for `combine_strs_with_missing_comments` to return `Result<String, RewriteError>`.

**Review by ytmimi — COMMENTED** [review]
Looks like we're on the right track with this. Left a few suggestions for way that we could improve things.

**Comment by ding-young** [comment]
@ytmimi Thank you for your review. It seems like I missed some details. I'll take a more careful look at the next PR. 
I am wondering to what extent I should modify functions like rewrite_*** in this PR. For example, you recommended modifying rewrite_let_else_block and rewrite_explicit_self, while you suggested changing the signature of combine_str_with_missing_comment later. Should I proceed with modifying the simpler ones for now?

**Comment by ytmimi** [comment]
I'm fairly certain that `rewrite_let_else_block` and `rewrite_explicit_self` are only used within these functions so it's likely easier to update their signatures now, while `combine_str_with_missing_comment` is used throughout the codebase and would require touching a lot of different files.

If you feel like it would be a lot to extend the scope of the current PR then we can always revisit modifying all of these functions in future PRs, but if it's not a heavy lift then we might want to consider it.

File: src/items.rs [hunk 5]
```diff
@@ -1892,25 +1911,33 @@ pub(crate) fn rewrite_struct_field(
...
+    let ty_shape = shape
         .offset_left(overhead + spacing.len())
-        .and_then(|ty_shape| field.ty.rewrite(context, ty_shape));
-    if let Some(ref ty) = orig_ty {
+        .max_width_error(shape.width, field.ty.span())?;
+    let orig_ty = field.ty.rewrite_result(context, ty_shape);
+    if let Ok(ref ty) = orig_ty {
         if !ty.contains('\n') && !contains_comment(context.snippet(missing_span)) {
-            return Some(attr_prefix + &spacing + ty);
+            return Ok(attr_prefix + &spacing + ty);
```
Comment (ytmimi) [inline src/items.rs:1925]: We need to rework this. We can't fail fast by using the try operator when calculating `ty_shape`. In the original code when we failed to calculate the `orig_ty` shape we wouldn't enter the `if let Some(ref ty) = orig_ty` block, and code execution would continue on to the code below. The current implementation short circuits if we fail to calculate the `ty_shape` which changes the behavior.

File: src/rewrite.rs [hunk 6]
```diff
@@ -44,7 +81,6 @@ pub(crate) struct RewriteContext<'a> {
     pub(crate) skip_context: SkipContext,
     pub(crate) skipped_range: Rc<RefCell<Vec<(usize, usize)>>>,
 }
-
```
Comment (ytmimi) [inline src/rewrite.rs:47]: Let's revert this change

Comment (ytmimi) [inline src/rewrite.rs:47]: Let's revert this whitespace change. It'll be better to keep a space to separate `RewriteContext` and `InsideMacroGuard`.

File: src/rewrite.rs [hunk 7]
```diff
@@ -13,9 +14,13 @@ use crate::skip::SkipContext;
 use crate::visitor::SnippetProvider;
 use crate::FormatReport;
 
+pub(crate) type RewriteResult = Result<String, RewriteError>;
 pub(crate) trait Rewrite {
     /// Rewrite self into shape.
     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String>;
+    fn rewrite_result(&self, context: &RewriteContext<'_>, shape: Shape) -> RewriteResult {
+        self.rewrite(context, shape).ok_or(RewriteError::Unknown)
+    }
```
Comment (ytmimi) [inline src/rewrite.rs:23]: nit: let's add a newline to separate `rewrite` and `rewrite_result`

File: src/items.rs [hunk 8]
```diff
@@ -1845,6 +1860,9 @@ pub(crate) fn rewrite_struct_field_prefix(
 
 impl Rewrite for ast::FieldDef {
     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String> {
+        self.rewrite_result(context, shape).ok()
+    }
+    fn rewrite_result(&self, context: &RewriteContext<'_>, shape: Shape) -> RewriteResult {
```
Comment (ytmimi) [inline src/items.rs:1865]: nit: Let's add a newline to separate `rewrite` and `rewrite_result`

File: src/items.rs [hunk 9]
```diff
@@ -2072,19 +2099,30 @@ impl<'a> Rewrite for OpaqueType<'a> {
 
 impl Rewrite for ast::FnRetTy {
     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String> {
+        self.rewrite_result(context, shape).ok()
+    }
+    fn rewrite_result(&self, context: &RewriteContext<'_>, shape: Shape) -> RewriteResult {
```
Comment (ytmimi) [inline src/items.rs:2104]: nit: let's add a newline to separate `rewrite` and `rewrite_result`

File: src/items.rs [hunk 10]
```diff
@@ -2072,19 +2099,30 @@ impl<'a> Rewrite for OpaqueType<'a> {
...
                         .map(|r| format!("-> {}", r));
                 }
 
-                ty.rewrite(context, shape.offset_left(3)?)
+                // 3 = "-> "
+                let shape = shape
+                    .offset_left(3)
+                    .max_width_error(shape.width, self.span())?;
+
+                ty.rewrite_result(context, shape)
```
Comment (ytmimi) [inline src/items.rs:2125]: nit: Since we're making changes here anyway, it might be nice to document all of the `+ 3` we've got. I'm pretty sure they all correspond to the length of `"-> "`. I know in a previous comment I suggested adding the comment `// 3 = "-> "`, but I think this would be an improvement on that.

```rust
                {
                    let arrow_width = "-> ".len();
                    let inner_width = shape
                        .width
                        .checked_sub(arrow_width)
                        .max_width_error(shape.width, self.span())?;
                    return ty
                        .rewrite_result(context, Shape::legacy(inner_width, shape.indent + arrow_width))
                        .map(|r| format!("-> {}", r));
                }

                let shape = shape
                    .offset_left(arrow_width)
                    .max_width_error(shape.width, self.span())?;

                ty.rewrite_result(context, shape)
```

File: src/items.rs [hunk 11]
```diff
@@ -73,15 +77,21 @@ impl Rewrite for ast::Local {
...
+            )
+            .unknown_error()?
         };
         let let_kw_offset = result.len() - "let ".len();
 
         // 4 = "let ".len()
-        let pat_shape = shape.offset_left(4)?;
+        let pat_shape = shape
+            .offset_left(4)
+            .max_width_error(shape.width, self.span())?;
```
Comment (ytmimi) [inline src/items.rs:88]: On thing I've been wondering is whether we should use `shap.width` or `context.config.max_width()` for the max_width_error. Would love to get your thoughts on this.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Thanks for working through all the feedback! Here's my follow up review

File: src/items.rs [hunk 12]
```diff
@@ -116,15 +128,16 @@ impl Rewrite for ast::Local {
 
         if let Some((init, else_block)) = self.kind.init_else_opt() {
             // 1 = trailing semicolon;
-            let nested_shape = shape.sub_width(1)?;
+            let nested_shape = shape.sub_width(1).unknown_error()?;
```
[the remark this answers is not in the document]
Comment (ding-young) [reply src/items.rs:131]: nope, this is my mistake.

File: src/items.rs (hunk 11 above)
Comment (ding-young) [reply src/items.rs:88]: I believe `shape.width` provides more specific information, considering that rewrite tries to fit the formatted code into the given shape, while `context.config.max_width()` is set by the global configuration. Even if we want to expose only `max_width` to users in error messages (since `shape.width` might be confusing), in most cases, callers can retrieve `context.config.max_width()` from its rewrite context so converting `RewriteError` to `ErrorKind` would be possible.

However, if there isn't much opportunity to retry formatting based on the returned `RewriteError` with `shape.width` and `span`, it might be pointless to include shape.width, especially if this information is unused.

File: src/items.rs [hunk 13]
```diff
@@ -1892,25 +1914,29 @@ pub(crate) fn rewrite_struct_field(
...
         .offset_left(overhead + spacing.len())
-        .and_then(|ty_shape| field.ty.rewrite(context, ty_shape));
+        .and_then(|ty_shape| field.ty.rewrite_result(context, ty_shape).ok());
+
     if let Some(ref ty) = orig_ty {
         if !ty.contains('\n') && !contains_comment(context.snippet(missing_span)) {
-            return Some(attr_prefix + &spacing + ty);
+            return Ok(attr_prefix + &spacing + ty);
         }
     }
```
Comment (ding-young) [inline src/items.rs:1926]: Since we are gonna continue formatting though `offset_left` returns None, I replaced `ty.rewrite` with `ty.rewrite_result.ok()` instead of converting Option into RwResult like below. 
```rust
let orig_ty = shape
        .offset_left(overhead + spacing.len())
        .max_width(shape.width, field.span())
        .and_then(|ty_shape| field.ty.rewrite_result(context, ty_shape));

    if let Ok(ref ty) = orig_ty {
        if !ty.contains('\n') && !contains_comment(context.snippet(missing_span)) {
            return Ok(attr_prefix + &spacing + ty);
        }
    }
```
Although it seems somewhat pointless to calling rewrite_result.ok() instead of rewrite, I did so since rewrite will return Result after all. 
It's pretty minor issue, but I want to hear your suggestion if you have one.

File: src/items.rs (hunk 11 above)
Comment (ytmimi) [reply src/items.rs:88]: I think it's fine to leave this as `shape.width` for now. We can always revisit this decision in the future if it turns out there's something better we should use instead of the width.

File: src/items.rs (hunk 13 above)
Comment (ytmimi) [reply src/items.rs:1926]: I think calling `rewrite_result.ok()` is fine. We could also follow your suggestion, but I don't think it makes much of a difference right now.

**Review by ytmimi — APPROVED** [review]
Great work. I took another look and I think we're good to go on this one

**Comment by ytmimi** [comment]
@ding-young when you have a chance can you squash the commits and rebase this PR to bring it up to date
